### ISSUE TITLE
Are tensorflow lite models that include select_tf_ops supported?

### ISSUE THREAD

**Issue by WalkingWithSpacemen3** [WasmEdge/WasmEdge#2398]
## Motivation

I have converted a TensorFlow Lite model containing [select_tf_ops](https://www.tensorflow.org/lite/guide/ops_select) and am attempting to run inference on this model using Wasmedge.

## Details

When attempting to run the inference process in Wasmedge, I receive the following error:  `ERROR: Select TensorFlow op(s), included in the given model, is(are) not supported by this interpreter. Make sure you apply/link the Flex delegate before inference. For the Android, it can be resolved by adding "org.tensorflow:tensorflow-lite-select-tf-ops" dependency. See instructions: https://www.tensorflow.org/lite/guide/ops_select` `[WASI-NN] Invokation failed.
thread 'main' panicked at 'Failed to execute inference: NnErrno { code: 3, name: "MISSING_MEMORY", message: "" }'`. I believe to use models containing select_tf_ops in a CMake project, you need to link the shared library [libtensorflowlite_flex.so](https://www.tensorflow.org/lite/guide/ops_select#cc) in the CMakeList.txt.
I've done some exploring in the Wasmedge-tensorflow repository, but am still unsure of what to do. Could anyone provide me with some guidance? Thank you!

**Comment by juntao** [comment]
I believe the issue is that you will need to use your custom tf library to replace the one in the WasmEdge NN plugin default install. 

@apepkuss What do you think?

**Comment by apepkuss** [comment]
Hi @WalkingWithSpacemen3 ,

Does your tf-lite model contain any custom operators? Or just contains tf-lite + tf operators? Thanks!

**Comment by WalkingWithSpacemen3** [comment]
> 
Hi, thanks for the reply! 
It doesn't contain any custom operators, just tf-lite and tf operators.

**Comment by apepkuss** [comment]
Hi @q82419, according to the [descriptions](https://wasmedge.org/book/en/contribute/build_from_src/plugin_wasi_nn.html#build-wasmedge-with-wasi-nn-tensorflow-lite-backend), the `wasi-nn tf-lite backend` is only integrated with `libtensorflowlite_c.so`. However, in some cases, the `tflite` models also contain the operators defined in `tf` (please refer to the description about the [ops compatibility](https://www.tensorflow.org/lite/guide/ops_compatibility)). To solve the issue, as @WalkingWithSpacemen3 mentioned above, the build script of the `wasi-nn tensorflow-lite backend` should add additional options. Tensorflow official webpage [Select TensorFlow operators](https://www.tensorflow.org/lite/guide/ops_select#convert_a_model) provides the solution to the issue. Could you please help solve the issue? Thanks a lot!

**Comment by WalkingWithSpacemen3** [comment]
To add some information, I encountered the same problem when using the official tensorflowlite example minimal to read my model. Then I compiled libtensorflowlite_flex.so, linked it and the problem was solved. Also found a quote about enabling select_tf_ops [here](https://www.tensorflow.org/lite/guide/signatures#from_saved_model): `TFLite interpreter needs to link Flex delegate in order to run the model since it contains the following Select TFop(s)`

**Comment by WalkingWithSpacemen3** [comment]
> Hi @WalkingWithSpacemen3 ,
> 
> Let me summarize this issue. You compiled the `libtensorflowlite_flex.so` with `libtensorflowlite_c.so` and link to them. So that you can use WasmEdge WASI-NN plug-in to run your example? If that, I think we can update the WasmEdge tensorflow dependencies with these pre-compiled shared libraries. Becuase our [guide](https://wasmedge.org/book/en/write_wasm/rust/wasinn.html#get-wasmedge-with-wasi-nn-plug-in-tensorflow-lite-backend) teach users to install the TFLite dependency from our manylinux version.
> 
> If the above is correct, can you help to provide your example and the guide to compile this shared library? Thanks!

Hi! 
Thanks for the reply.
No, the actual case is, I compiled the tensorflowlite library and I wanted to test my compiled library and my model with the official [minimal](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/examples/minimal) example. Before I link libtensorflowlite_flex.so, I get the same error as described above when I run the minimal example, and when I link the flex library this way `find_library(TF_LIB_FLEX tensorflowlite_flex HINTS "/PATH/TO/LIB")
target_link_libraries(minimal
  -Wl,--no-as-needed
tensorflow-lite
${TF_LIB_FLEX}
)`, the minimal example runs fine. But the problem is I don't know where to link the tensorflowlite flex library in WasmEdge.

**Comment by q82419** [comment]
> 

Hi, 

The `WASI-NN` plug-in with TFLite backend of WasmEdge linked the `libtensorflowlite_c.so` with accessing the TFLite C API.
I think that as the example you mentioned and modified, you link the `libtensorflowlite_flex.so` additionally.
Can you provide the `ldd` result of that binary?

And also, we need the example for using WASI-NN plug-in in WasmEdge, especially the WASM file.
Therefore we can test and update the plug-in and dependency.

Thanks!

**Comment by q82419** [comment]
> Hi,
> 
> following is the output of `ldd libtensorflowlite_flex.so`: `linux-vdso.so.1 (0x00007fffccff1000)`, `libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fd9cfca8000)`, `libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fd9cfa7e000)`, `libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fd9cfa5e000)`, `libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fd9cf836000)`, `/lib64/ld-linux-x86-64.so.2 (0x00007fd9daeb7000)`.
> 
> I put my codes [here](https://github.com/WalkingWithSpacemen3/Stable-Diffusion-Wasi-NN), the wasm file and the binaries I built can be found in the release section. My tflite models can also be downloaded from [here](https://drive.google.com/file/d/1MirVHIKV7uGdBcH2wnwUSpsql0EeIUkX/view?usp=share_link) in case they are needed.
> 
> Thanks!

Sorry that I didn't describe clearly.
Can you provide the `ldd` of the executable compiled from the official example you modified?
That is, the official [minimal](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/examples/minimal) example with the flex library you linked.

**Comment by WalkingWithSpacemen3** [comment]
oh, then the output will be `linux-vdso.so.1 (0x00007ffe0b0f2000)`, `libtensorflowlite_flex.so => /usr/local/lib/libtensorflowlite_flex.so (0x00007fd3d4702000)`, `libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fd3d4616000)`, `libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fd3d43ec000)`, `libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fd3d43cc000)`, `libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fd3d41a4000)`, `/lib64/ld-linux-x86-64.so.2 (0x00007fd3dfd66000)` .

**Comment by q82419** [comment]
> oh, then the output will be `linux-vdso.so.1 (0x00007ffe0b0f2000)`, `libtensorflowlite_flex.so => /usr/local/lib/libtensorflowlite_flex.so (0x00007fd3d4702000)`, `libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fd3d4616000)`, `libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fd3d43ec000)`, `libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fd3d43cc000)`, `libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fd3d41a4000)`, `/lib64/ld-linux-x86-64.so.2 (0x00007fd3dfd66000)` .

Because we use TFLite C API now.
Maybe we should spend time to try to select the TF Ops with C API.
Thanks.

**Comment by yanghaku** [comment]
Hello @WalkingWithSpacemen3 , I have tested the WasmEdge tflite plugin with select TF ops and run your project successfully.

## Steps

1. First, use the **TensorFlow >= v2.7**, because the feature enables Flex delegate by linking a TensorFlow Lite Flex delegate shared library [can only be available since version v2.7](https://www.tensorflow.org/lite/guide/ops_select#cc).
2. Then, build the WasmEdge and link with your ```libtensorflowlite_flex.so``` and ```libtensorflowlite_c.so```  (version >= 2.7). 
* for the  ```libtensorflowlite_c.so``` , you must replace the file in ```_deps/wasmedgetensorflowdepslite-src```
* for the ```libtensorflowlite_flex.so```, you can use [this branch to build](https://github.com/yanghaku/WasmEdge/tree/tf_select_test) and just set the CMake options ```WASMEDGE_TENSORFLOW_DEPS_TFLITE_FLEX_PATH```.

## Details

### The running result in the console

```console
Read text encoder weights, size in bytes: 246469972
Read diffusion model weights, size in bytes: 1720482640
Read decoder weights, size in bytes: 99171900
Loaded graph into wasi-nn with ID: 0
Loaded graph into wasi-nn with ID: 1
Loaded graph into wasi-nn with ID: 2
Created text encoder execution context with ID: 0
INFO: Created TensorFlow Lite delegate for select TF ops.
2023-04-25 19:29:44.929263: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE3 SSE4.1 SSE4.2 AVX AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
INFO: TfLiteFlexDelegate delegate: 61 nodes delegated out of 5909 nodes with 61 partitions.

Created diffusion model execution context with ID: 1
Executed text encoder inference
Executed text encoder inference
diffusion model inference start point
Executed diffusion model inference

......
```

Additionally, your output buffer is too small in src/main.rs Line 239.


### patch to link the  ```libtensorflowlite_flex.so``

```diff
diff --git a/plugins/wasi_nn/CMakeLists.txt b/plugins/wasi_nn/CMakeLists.txt
index 74517d89..ad319e10 100644
--- a/plugins/wasi_nn/CMakeLists.txt
+++ b/plugins/wasi_nn/CMakeLists.txt
@@ -154,6 +154,12 @@ foreach(BACKEND ${WASMEDGE_PLUGIN_WASI_NN_BACKEND})
       PUBLIC
       ${WASMEDGE_TENSORFLOW_DEPS_TFLITE_LIB}
     )
+    if(DEFINED WASMEDGE_TENSORFLOW_DEPS_TFLITE_FLEX_PATH)
+      target_link_libraries(wasmedgePluginWasiNN
+        PUBLIC
+        ${WASMEDGE_TENSORFLOW_DEPS_TFLITE_FLEX_PATH}
+      )
+    endif()
   else()
     # Add the other backends here.
     message(FATAL_ERROR "WASI-NN: backend ${BACKEND} not found or unimplemented.")
```

**Comment by WalkingWithSpacemen3** [comment]
Thanks! I'll close this issue then.

**Comment by q82419** [comment]
> Hello @WalkingWithSpacemen3 , I have tested the WasmEdge tflite plugin with select TF ops and run your project successfully.
> 
> ## Steps
> 1. First, use the **TensorFlow >= v2.7**, because the feature enables Flex delegate by linking a TensorFlow Lite Flex delegate shared library [can only be available since version v2.7](https://www.tensorflow.org/lite/guide/ops_select#cc).
> 2. Then, build the WasmEdge and link with your `libtensorflowlite_flex.so` and `libtensorflowlite_c.so`  (version >= 2.7).
> 
> * for the  `libtensorflowlite_c.so` , you must replace the file in `_deps/wasmedgetensorflowdepslite-src`
> * for the `libtensorflowlite_flex.so`, you can use [this branch to build](https://github.com/yanghaku/WasmEdge/tree/tf_select_test) and just set the CMake options `WASMEDGE_TENSORFLOW_DEPS_TFLITE_FLEX_PATH`.
> 
> ## Details
> ### The running result in the console
> ```
> Read text encoder weights, size in bytes: 246469972
> Read diffusion model weights, size in bytes: 1720482640
> Read decoder weights, size in bytes: 99171900
> Loaded graph into wasi-nn with ID: 0
> Loaded graph into wasi-nn with ID: 1
> Loaded graph into wasi-nn with ID: 2
> Created text encoder execution context with ID: 0
> INFO: Created TensorFlow Lite delegate for select TF ops.
> 2023-04-25 19:29:44.929263: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE3 SSE4.1 SSE4.2 AVX AVX2 FMA
> To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
> INFO: TfLiteFlexDelegate delegate: 61 nodes delegated out of 5909 nodes with 61 partitions.
> 
> Created diffusion model execution context with ID: 1
> Executed text encoder inference
> Executed text encoder inference
> diffusion model inference start point
> Executed diffusion model inference
> 
> ......
> ```
> 
> Additionally, your output buffer is too small in src/main.rs Line 239.
> 
> ### patch to link the ```libtensorflowlite_flex.so``
> ```diff
> diff --git a/plugins/wasi_nn/CMakeLists.txt b/plugins/wasi_nn/CMakeLists.txt
> index 74517d89..ad319e10 100644
> --- a/plugins/wasi_nn/CMakeLists.txt
> +++ b/plugins/wasi_nn/CMakeLists.txt
> @@ -154,6 +154,12 @@ foreach(BACKEND ${WASMEDGE_PLUGIN_WASI_NN_BACKEND})
>        PUBLIC
>        ${WASMEDGE_TENSORFLOW_DEPS_TFLITE_LIB}
>      )
> +    if(DEFINED WASMEDGE_TENSORFLOW_DEPS_TFLITE_FLEX_PATH)
> +      target_link_libraries(wasmedgePluginWasiNN
> +        PUBLIC
> +        ${WASMEDGE_TENSORFLOW_DEPS_TFLITE_FLEX_PATH}
> +      )
> +    endif()
>    else()
>      # Add the other backends here.
>      message(FATAL_ERROR "WASI-NN: backend ${BACKEND} not found or unimplemented.")
> ```

Thanks for help.
We'll discuss this for the `0.13.0` release in the future.
